modal 컴포넌트 백그라운드 컴포넌트 생성

lacomaco opened the pull request:
## 관련 이슈
[공통 컴포넌트] 모달 컴포넌트 #87 

## 구현/수정 내용
- 자신이 포함하고 있지 않은 DOM이 클릭되어졌을때 발동하는 useClickOutside 커스텀 훅을 생성했습니다.
- 자기 자신이 클릭되어졌을때 이벤트를 발생시키는 useClickMe 커스텀훅을 생성했습니다.
- modal의 검정 백그라운드 컴포넌트를 생성했습니다.
- modal의 형태와 modal이 기능하는 예제를 담은 스토리를 생성했습니다.

@boostcamp-2020/accountbook_mentor

lacomaco committed:
feat : #87 모달 컴포넌트의 기능 customHook 구현

- 자신 이외의 것이 클릭되면 callback인자를 호출하는 customHook을
  추가하였습니다.

- 자신 이외의 것이 클릭되면 모달이 꺼지게 하는 용도로 사용하기 위해서
  추가했어요

lacomaco committed:
feat : #87 useClickOutside CustomHook 테스트 코드 작성

- 자신의 안쪽이 클릭되어졌을 경우와
- 바깥쪽 인자가 클릭되어졌을 경우에 대한 테스트 코드를 추가하였습니다.

lacomaco committed:
feat : #87 자기 자신만 클릭되어졌을때 동작하는 customHooks 생성

- useClickMe는 정확히 자기 자신만 클릭되어졋을때 동작하는
  customHooks입니다.

- useClickMe는 이벤트 버블링을 무시합니다.

- useClickMe는 첫번째 인자로 자기 자신의 ref을 넘겨주고 두번째 인자로
  수행할 콜백 함수를 받고 클릭 되어졋을때 콜백을 실행합니다.

- modal에서 검정 백그라운드가 눌려졌을때 닫기 위해서 만들어둔 함수에요

lacomaco committed:
feat : #87 stories에 modal background가 켜져 있는 부분을 추가

lacomaco committed:
feat : #87 modalBackground 컴포넌트 생성

- modalBackground 컴포넌트는 children을 modal의 정중앙에 그려주는
  컴포넌트입니다.

- props로 show 상태와 모달을 닫는 closeModal 함수를 받습니다.

- modalBackground를 사용하기 위해서는 modal 상태를 관리하는 container
  컴포넌트로 감싸준후 사용해야합니다.

lacomaco committed:
feat : #87 모달 컴포넌트에 사용되는 useClickMe 테스트 코드 추가

- 자신 이외의 컴포넌트가 클릭되었을 경우에 대한 테스트 코드 추가
- 자기 자신이 클릭되어졌을 경우에 대한 테스트 코드 추가

lacomaco committed:
feat : #87 useClickMe 테스트 강화

- 자식 컴포넌트가 눌려졌을 경우에 반응하지 않는다는 테스트 케이스 추가

mu1616 commented on client/src/hook/use-click-outside/useClickOutside.tsx:
```diff
@@ -0,0 +1,20 @@
+  useEffect(() => {
+    const handleClickOutSide = (e: MouseEvent): void => {
+      e.stopPropagation();
+      if (ref.current && !ref.current.contains(e.target as Node)) {
```
e.target as Node 처음 보는 문법인데 무슨 뜻인가요..??

mu1616 replied:
ㅎㅇ

mu1616 commented on client/src/components/modal-background/ModalBackground.stories.tsx:
```diff
@@ -0,0 +1,36 @@
+  return (
+    <div>
+      <button onClick={openModal}>모달을 열어봅시다</button>
+      <ModalBackground show={show} closeModal={closeModal}>
+        <div>안녕하세요</div>
+      </ModalBackground>
```
혹시 이 부분을 display:none으로 조작하는게 아니라 아래와 같이 show 상태에 따라 렌더링 여부를 결정해주는건 어떻게 생각하시나요??
```
{show && <ModalBackground show={show} closeModal={closeModal}>
        <div>안녕하세요</div>
      </ModalBackground>
```

mu1616 reviewed (approved):
커스텀훅에 테스트코드까지.. 항상 영근님 코드보고 많이 배웁니다!!
수고 많으셨습니다!! :smile:

Mong-Gu reviewed (approved):
솔직히 완전 다 이해는 못했지만 하면서 더 참고하도록 하겠습니다..! 수고하셨습니다!

bbbyung2 commented on client/src/components/modal-background/ModalBackground.stories.tsx:
```diff
@@ -0,0 +1,36 @@
+    <ModalBackground
+      show={true}
+      closeModal={(): void => {
+        1 + 1;
```
1 + 1 수정해주시면 좋을 것 같아요!

lacomaco replied:
수정했습니다 🥰

lacomaco committed:
fix : #87 close 함수 변경

- close 함수가 props로 넘어오지 않아도 렌더링 되도록 변경
- close 함수가 있을 경우에만 useClickMe 커스텀 훅을 바인딩해서 닫히도록
  변경
- storybook의 의미없는 1+1 제거
